### PR TITLE
NEW: Move basic inference workflow into installable python module

### PR DESCRIPTION
**Pull request by carterbox**
The purpose of this PR is to create an installable python package with various PtychoNN workflows available as functions. Having an installable python package makes reuse of existing code easier by other workflows and allows distribution via pip or conda.

This first PR only creates the framework for the package and implements the inference workflow using @vbanakha's small NN model as a starting point.

It also creates a ptychonn CLI based on @vbanakha's workflows, but I'm not sure that these CLI will be kept in the long-run. Mostly I have been just using them for testing purposes. I think that @stevehenke, will be using the function APIs instead.

This PR merges into a new "package" branch because I have been deleting files from my branch. The files are still in the git history, but I don't want to make any of the legacy scripts hard to find at this point in time.

@vbanakha, please review.

### CODE REVIEW

**Comment by carterbox** [comment]
I moved most of the deletions of the research code from the master branch onto the package branch, so this PR only shows the new stuff (is easier to read). The deleted files still exist in the history/on the other branch, so they are not lost.

File: ptychonn/_infer/__main__.py [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+            'ptychonn._infer',
+            'weights.pth',
+    ) as load_model_path:
+        click.echo(f'Loading model at {load_model_path}')
+        recon_model = helper_small_model.ReconSmallPhaseModel()
+        tester = helper_small_model.Tester(
+            model=recon_model,
+            batch_size=max(torch.cuda.device_count(), 1) * 64,
+            model_params_path=load_model_path,
+        )
```
Comment (carterbox) [inline ptychonn/_infer/__main__.py]: @stevehenke requested that the ptychonn.infer() API allow arbitrary model inputs, but remain a datastructure (not file-based API), so I need to move the instantiation of this "Tester" class outside of this function.

File: ptychonn/_infer/__main__.py [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+    spiral_step = 0.05
+    step = spiral_step * -1e-6
+    spiral_traj = np.load(scan_path)
+    scan = np.stack((spiral_traj['x'], spiral_traj['y']), axis=-1) * step
+    stitched = stitch_from_inference(
+        inferences,
+        scan,
+        stitched_pixel_width=10e-9,
+        inference_pixel_width=10e-9,
+    )
```
Comment (carterbox) [inline ptychonn/_infer/__main__.py:137]: Start here to understand the workflow. At a high level, the workflow steps are:

1. Load diffraction patterns
2. Give the diffraction patterns to the trained network for inference
3. Stitch the network output together into a full reconstructed image

File: ptychonn/_infer/__main__.py [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+    inference_pixel_width: float,
+    pix: int = 0,
+) -> npt.NDArray:
+    '''Combine many overlapping inferred patches into a single image.
+
+    Parameters
+    ----------
+    inferences : (POSITION, WIDTH, HEIGHT)
+        Overlapping patches of the inferred field of view.
+    scan : (POSITION, 2)
```
Comment (carterbox) [inline ptychonn/_infer/__main__.py:30]: This should have units or be marked as unitless.

File: ptychonn/_infer/__main__.py [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+        The relative coordinates of each of the overlapping patches.
```
Comment (carterbox) [inline ptychonn/_infer/__main__.py]:
```suggestion
    scan : (POSITION, 2) [m]
        The coordinates of each of the overlapping patches.
```

**Review by vbanakha — APPROVED** [review]
The functionality seems fine.

**Pull request merged**